crates/wasmparser/src/validator/operators.rs

```diff
@@ -3442,6 +3473,79 @@ where
         self.pop_operand(Some(ValType::I32))?;
         Ok(())
     }
+    fn visit_struct_new_default(&mut self, type_index: u32) -> Self::Output {
+        let ty = self.struct_type_at(type_index)?;
+        for i in 0..u32::try_from(ty.len_fields()).unwrap() {
+            let field = ty.field_at(i).unwrap();
+            let val_ty = field.element_type.unpack();
+            if !val_ty.is_defaultable() {
+                bail!(
+                    self.offset,
+                    "invalid `struct.new_default`: {val_ty} field is not defaultable"
+                );
+            }
+        }
+
+        let mut heap_ty = HeapType::Concrete(UnpackedIndex::Module(type_index));
+        self.resources.check_heap_type(&mut heap_ty, self.offset)?;
```

Comment:
I think this can be skipped since `struct_type_at` above would have otherwise failed?

Reply:
This is to canonicalize the module index into an id. I'll add a comment to this effect.